### ISSUE TITLE
Prompt user to convert .HEIC files to .png/.jpg

### ISSUE THREAD

**Issue by MMachado05** [localsend/localsend#949]
### Is your feature request related to a problem? Please describe.

iPhones default to the HEIC filetype when taking pictures, and many users don't bother switching off this setting as it does play nicely with the rest of the iPhone's capabilities. Unfortunately, this has yet to become a totally adopted filetype. On my Linux system, for example, my image viewer Gwenview cannot open this kind of format. So, when I send photos to my laptop, I have to painstakingly convert all of the HEIC files to JPGs, usually by opening them in GIMP and just immediately exporting them.

### Describe the solution you'd like

To attack this generally, since there may be other troublesome filetypes I'm unfamiliar with, I think that LocalSend should search the uploaded files for any formats that might not be as cross-platform as others, then prompt the user on if they want them to be converted before sending. Something like,

"The following files may be unsupported on other systems. (then a list). Would you like to convert them before sending? (here, if there are multiple unsupported filetypes, there could be a list like HEIC → JPG and others if relevant.) [] Don't ask again"

### Describe alternatives you've considered

Either that, or just do the conversion immediately, and just have a "map filetypes" setting in the settings so that users have more control. More complex, yes, but for many users who might not care, it gets out of their way and becomes an "advanced setting."

### Additional context

N/A

**Comment by GarryOneW00d** [comment]
I have all Apple devices, so I haven't encountered such a problem. But checking files for compatibility and conversion seems a very strange implementation, in an application that is intended for file transfer, especially given the large number of unresolved issues that directly concern the purpose of this app.

**Comment by EdwardWilsonJr** [comment]
I agree with @GarryOneW00d, this is a completely unnecessary feature for a file transfer app. Because this includes any format that isn't supported by the system by default, even the same doc or docx that requires the Office package. I'm generally silent about commercial closed formats. And you can transfer a file or files for various reasons, and it isn't always in order to view it, but simply, for example, for a backup copy.
As for HEIC support, for Linux it's solved using the 'heif‑gdk‑pixbuf' package. All new versions of Windows support HEIC out of the box, through the standard Photos app. Therefore, such implementation is a waste of time. Better to really focus on the visible problems at hand.

**Comment by Tienisto** [comment]
Yes, there are more important things to do.

Converting HEIC to JPEG would also decrease the quality of the image which doesn't fit to a file sharing app.

**Comment by MMachado05** [comment]
@EdwardWilsonJr I appreciate this discussion, and it's made me rethink my original proposal. It should be on the receving device to ensure proper file support rather than on a third-party application to play interpreter; I think this should be labeled as an "upstream issue" of sorts, considering this issue would be moot if operating systems just shipped with the necessary file support out-of-the-box.

@Tienisto, is there a way to close this as an "upstream bug"? Just to denote that the functionality is out of scope for the project. That is, of course, assuming you agree; as the primary maintainer of the project, if this is still something you decide to implement, then doing so would be unnecessary.

**Comment by EdwardWilsonJr** [comment]
> Yes, there are more important things to do.
> 
> Converting HEIC to JPEG would also decrease the quality of the image which doesn't fit to a file transfer app.
> 
> However, I have noticed some feedback that users could not open HEIC images on their receiving device. A warning box would certainly help non-technical users. Providing _optional_ conversion from HEIC to JPEG may be possible to ensure minimal effort by the user.

HEIC uses its own color profile and color space, so it's not just about loss in quality.

As for the warning about potential format incompatibility and the suggestion to convert images to another format. I doubt that an average user would know how to do that. Embedding the ability to convert images directly into the application sounds like complete nonsense. How would it differ from using ProRes for video, which is also not supported by all operating systems? Not to mention that most distributions, in their majority, use a fully open-source model, so most popular video codecs are not included out of the box and need to be installed. Let's embed a video converter then. Windows, for example, don't have a built-in archiver until recently, and older versions still don't have it. Let's throw an archiver in here to open RAR archives. It's some kind of absurdity. Of course, it's up to you to decide, but opening such a Pandora's box that will only cause problems and is completely unrelated to the application's direction is just a ridiculous idea. Even the author of the request agreed that it looks strange from the outside. And the fact that none of similar applications supports this should probably be a cause for concern.

**Comment by EdwardWilsonJr** [comment]
> @EdwardWilsonJr I appreciate this discussion, and it's made me rethink my original proposal. It should be on the receving device to ensure proper file support rather than on a third-party application to play interpreter; I think this should be labeled as an "upstream issue" of sorts, considering this issue would be moot if operating systems just shipped with the necessary file support out-of-the-box.

I'm glad that you were able to take a look at your request objectively and see that it doesn't directly relate to LocalSend. It would be more appropriate to address it to the developers of Gwenview. Since Gwenview is part of the KDE suite of programs, developed and maintained by KDE themselves, they have the resources to make improvements. In general, KDE is about a hundred thousand different toggles and settings, so it's surprising that their application still doesn't support HEIC, for example through a plugin in the settings.

**Comment by sergd88** [comment]
I read the comments, in general, some strange discussion arose. It seems logical that the purpose of LocalSend is "Share files to nearby devices". It even says so on the website. Its function is to transfer the file/s intact and undamaged. If the transfer of the file/s was successful, but the file/s don't open for some reason, this isn't a matter for LocalSend. I don't know what else can be discussed here.

**Comment by 75th** [comment]
The "strange discussion" and "absurdity" here is on the part of the developers, not the OP. LocalSend is most likely to be used between platforms not from the same vendor. If you have all Apple devices, you use AirDrop; LocalSend is for people with the need to send between heterogeneous devices, which may not have equivalent file format support.

In particular, LocalSend for iOS has the ability to send files directly from the system photo library, which iOS provides by default as HEIC files. It's not "absurd" for an app about cross-platform file transfers to make files more compatible for the receiving device, *especially* when iOS has [native APIs](https://developer.apple.com/documentation/uikit/uiimage/jpegdata\(compressionquality:\)) for easily doing the conversion.

If there are other priorities that's fine, it's of course the developer's prerogative to put what they want in their app, but to treat this as an unreasonable request is laughable and offensive.

**Comment by melcurve** [comment]
Adding an "Auto Convert HEIC to JPG" option in LocalSend's settings is a practical idea—not "absurd" at all. When uploading a .HEIC file in Safari, it automatically converts HEIC images to JPG, just like many other iOS apps do when selecting photos from the album. Although LocalSend feels more like an app made by geeks, incorporating some user-friendly design is completely reasonable.

**Comment by Punteillos** [comment]
This feature will be fenomenal, i think its a problem when you send a heic file to windows and this is converted to jpg without ask.

**Comment by jeslor** [comment]
why not use this npm package which does all the conversion, compression and cropping
https://www.npmjs.com/package/crop-image-pro

**Comment by Mikann20041029** [comment]
Yeah, HEIC is such a pain on Linux/Windows. If you just want a quick workaround (no uploads), I made a tiny client-side page that batch converts HEIC→JPG/PNG in the browser and downloads a ZIP:
https://mikann20041029.github.io/local-file-tools/

Might help until native HEIC handling is improved.

**Comment by jiaton** [comment]
I'd like to support the OP's perspective. While I understand the maintainer's concern about "feature creep" and "image purity," we have to acknowledge a real-world pain point: the "all-Apple environment" struggle.

For many of us, LocalSend is the last line of defense for staying on the Android/Linux platform. In a social circle full of iPhones, getting HEIC files that lag on Android is a daily frustration. We don't always need the "source of truth"; sometimes we just need "it just works" usability.

If this doesn't fit the main branch's philosophy, perhaps a community-maintained fork or a plugin architecture is the way to go. Even if it adds some bloat, having a toggle to "Auto-convert to JPG on receive" would be a lifesaver. Without these user-centric bridges, the friction of staying outside the Apple ecosystem becomes almost unbearable.

**Comment by jiaton** [comment]
I just created a fork and implemented for Android devices. Github Action Artifacts APK (not signed) 
https://github.com/jiaton/localsend/actions/runs/24820869927

It's in the settings, toggle on `Auto-convert HEIC to JPG`. Works regardless whether `Save media to galley` is on. 
It actually has a one click hotspot feature btw when I implement above. You can use IOS to scan the QR code to join the Android device's hotspot and transfer photos.
